File: src/problemset/doublePointer/LongestSubstringWithoutRepeatingCharacters.go
package main

// https://leetcode-cn.com/problems/longest-substring-without-repeating-characters/submissions/
// time: 16ms, 31.03%
// mem: 3.1MB, 45.16%
func lengthOfLongestSubstring(s string) int {
	p, q := 0, 0
	m := 0
	set := map[byte]int{}

	/* abcabcbb
	   p q
	*/
	for q < len(s) {
		if set[s[q]] == 0 {
			set[s[q]]++
			q++
		} else {
			for s[p] != s[q] {
				set[s[p]]--
				p++
			}
			set[s[p]]--
			p++
		}
		m = max(m, q-p)
	}

	return m
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

File: src/problemset/doublePointer/LongestSubstringWithoutRepeatingCharacters.java
package problemset.doublePointer;

/*
 * @date_time: 2020-05-02 11:58
 * @author: MoeexT
 * @problem: 3
 * @url: https://leetcode-cn.com/problems/longest-substring-without-repeating-characters/
 */

import java.util.HashMap;
import java.util.Map;

public class LongestSubstringWithoutRepeatingCharacters {
    public int lengthOfLongestSubstring(String s) {
        if (s.length() == 0) return 0;

        int max = 0, left = 0;
        Map<Character, Integer> map = new HashMap<>();

        for (int i = 0; i < s.length(); i++) {
            char ch = s.charAt(i);
            if (map.containsKey(ch)) {
                left = Math.max(left, map.get(ch) + 1);
            }
            map.put(ch, i);
            max = Math.max(max, i - left + 1);
        }
        return max;
    }

    public static void main(String[] args) {
        LongestSubstringWithoutRepeatingCharacters solution = new LongestSubstringWithoutRepeatingCharacters();
        System.out.println(solution.lengthOfLongestSubstring("vars"));
    }
}

